src/flp/types/fixedpoint_l2.rs

```diff
@@ -0,0 +1,816 @@
+
+        // In order to compare the actual norm of the vector with the claimed
+        // norm, the field needs to be able to represent all numbers that can
+        // occur during the computation of the norm of any submitted vector,
+        // even if its norm is not bounded by 1. Because of our encoding, an
+        // upper bound to that value is `entries * 2^(2*bits - 2)` (see docs of
+        // compute_norm_of_entries for details). It has to fit into the field.
+        let err = Err(FlpError::Encode(format!(
+            "number of entries ({}) not compatible with field size",
+            entries,
+        )));
+
+        if let Some(val) = (entries as u128).checked_mul(1 << bits_for_norm) {
+            if let Ok(modulus) = u128::try_from(F::modulus()) {
+                if val >= modulus {
+                    return err;
+                }
+            } else {
+                return err;
+            }
+        } else {
+            return err;
+        }
+
+        // Construct the polynomial that computes a part of the norm for a
+        // single vector entry.
+        //
+        // the linear part is 2^n,
+        // the constant part is 2^(2n-2),
+        // the polynomial is:
+        //   p(y) = 2^(2n-2) + -(2^n) * y + 1 * y^2
+        let linear_part = fi_one << (fi_bits_per_entry);
+        let constant_part = fi_one << (fi_bits_per_entry + fi_bits_per_entry - fi_two);
+        let norm_summand_poly = vec![F::from(constant_part), -F::from(linear_part), F::one()];
+
+        // Compute chunk length and number of calls for parallel sum gadgets.
+        let len0 = bits_per_entry * entries + bits_for_norm;
+        let gadget0_chunk_len = std::cmp::max(1, (len0 as f64).sqrt() as usize);
+        let mut gadget0_calls = len0 / gadget0_chunk_len;
+        if len0 % gadget0_chunk_len != 0 {
+            gadget0_calls += 1;
+        }
+
+        let len1 = entries;
+        let gadget1_chunk_len = std::cmp::max(1, (len1 as f64).sqrt() as usize);
+        let mut gadget1_calls = len1 / gadget1_chunk_len;
+        if len1 % gadget1_chunk_len != 1 {
+            gadget1_calls += 1;
+        }
+
+        Ok(Self {
+            bits_per_entry,
+            fi_bits_per_entry,
+            entries,
+            bits_for_norm,
+            range_01_checker: poly_range_check(0, 2),
+            norm_summand_poly,
+            phantom: PhantomData,
+
+            // range constants
+            range_norm_begin: entries * bits_per_entry,
+            range_norm_end: entries * bits_per_entry + bits_for_norm,
+
+            // configuration of parallel sum gadgets
+            gadget0_calls,
+            gadget0_chunk_len,
+            gadget1_calls,
+            gadget1_chunk_len,
+        })
+    }
+}
+
+impl<
+        T: Fixed,
+        F: FieldElement,
+        SPoly: ParallelSumGadget<F, PolyEval<F>>,
+        SBlindPoly: ParallelSumGadget<F, BlindPolyEval<F>>,
+    > Clone for FixedPointBoundedL2VecSum<T, F, SPoly, SBlindPoly>
```

Comment:
Is it possible to derive this automatically? Try adding "Clone" to "#[derive(Debug, PartialEq, Eq)]" above the type declaration above.

Reply:
`SPoly` and `SBlindPoly` would need to be `Clone`able for that :(

Reply:
No, unfortunately this does not work, because the `ParallelSumGadget`s are not clonable. In our type they only appear in as phantom data, but it seems that rust does not take that into account:

```
error[E0277]: the trait bound `SPoly: Clone` is not satisfied
   --> src/flp/types/fixedpoint_l2.rs:357:31
    |
357 | impl<T, F, SPoly, SBlindPoly> Type for FixedPointBoundedL2VecSum<T, F, SPoly, SBlindPoly>
    |                               ^^^^ the trait `Clone` is not implemented for `SPoly`
    |
note: required because of the requirements on the impl of `Clone` for `FixedPointBoundedL2VecSum<T, F, SPoly, SBlindPoly>`
   --> src/flp/types/fixedpoint_l2.rs:179:32
    |
179 | #[derive(Debug, PartialEq, Eq, Clone)]
    |                                ^^^^^
```

Reply:
Indeed, the problem here is that the compiler doesn't know if the instantiation of ParallelSumGadget implements Clone. This fix here is tell the compiler that it must by refining the trait bounds. Here's one solution: https://github.com/divviup/libprio-rs/commit/b212dcf8a4744f3665492796cb81563dc425fcab

Reply:
Ah, thanks. I did that now.